File: references/2-示例代码/Python/SDK工具类/joinpay_signature.py
# -*- coding: utf-8 -*-
"""
汇聚支付 MD5 签名工具

签名算法（与 SignBiz.java 一致）：
1. 排除 hmac 参数
2. 按 key 字母升序排序
3. 拼接所有 value（不含 key）
4. MD5(拼接值 + merchantKey).toUpperCase()
"""

import hashlib


def sign(params: dict, merchant_key: str) -> str:
    """生成MD5签名

    Args:
        params: 请求参数字典（不包含hmac）
        merchant_key: 商户密钥

    Returns:
        签名值（大写）
    """
    # 排除hmac，按key排序
    filtered = {}
    for k, v in params.items():
        if k.lower() == "hmac":
            continue
        val = str(v).strip() if v is not None else ""
        if val != "":
            filtered[k] = val

    sorted_keys = sorted(filtered.keys())
    sign_str = "".join(filtered[k] for k in sorted_keys)
    return hashlib.md5((sign_str + merchant_key).encode("utf-8")).hexdigest().upper()


def verify(params: dict, merchant_key: str) -> bool:
    """验证MD5签名

    Args:
        params: 包含hmac的完整参数字典
        merchant_key: 商户密钥

    Returns:
        True=验签通过
    """
    received_hmac = str(params.get("hmac", "")).upper()
    calculated_hmac = sign(params, merchant_key)
    return calculated_hmac == received_hmac


def build_sign_string(params: dict) -> str:
    """构建待签名字符串（用于调试）"""
    filtered = {k: v for k, v in params.items() if k.lower() != "hmac" and v}
    sorted_keys = sorted(filtered.keys())
    return "".join(str(filtered[k]).strip() for k in sorted_keys)

File: references/2-示例代码/PHP/SDK工具类/JoinPaySignature.php
<?php
/**
 * 汇聚支付 MD5 签名工具
 *
 * 签名算法（与 SignBiz.java 一致）：
 * 1. 排除 hmac 参数
 * 2. 按 key 字母升序排序
 * 3. 拼接所有 value（不含 key）
 * 4. MD5(拼接值 + merchantKey).toUpperCase()
 */
class JoinPaySignature
{
    /**
     * 生成MD5签名（与 SignBiz.java:80-99 一致，空值不参与签名）
     */
    public static function sign(array $params, string $merchantKey): string
    {
        // 排除hmac，过滤空值
        $filtered = [];
        foreach ($params as $k => $v) {
            if (strtolower($k) === 'hmac') {
                continue;
            }
            $val = trim(strval($v ?? ''));
            if ($val !== '') {
                $filtered[$k] = $val;
            }
        }

        // 按key字母排序
        ksort($filtered);

        // 拼接value
        $signStr = implode('', $filtered);

        // MD5(拼接值 + 密钥).toUpperCase()
        return strtoupper(md5($signStr . $merchantKey));
    }

    /**
     * 验证MD5签名
     *
     * @param array $params 包含hmac的完整参数
     * @param string $merchantKey 商户密钥
     * @return bool 验签是否通过
     */
    public static function verify(array $params, string $merchantKey): bool
    {
        $receivedHmac = strtoupper(strval($params['hmac'] ?? ''));
        $calculatedHmac = self::sign($params, $merchantKey);
        return hash_equals($calculatedHmac, $receivedHmac);
    }

    /**
     * 构建待签名字符串（调试用）
     */
    public static function buildSignString(array $params): string
    {
        $filtered = array_filter($params, fn($k) => strtolower($k) !== 'hmac', ARRAY_FILTER_USE_KEY);
        $filtered = array_filter($filtered, fn($v) => trim(strval($v)) !== '');
        ksort($filtered);
        return implode('', $filtered);
    }
}
